src/firebase/api/gcp-utils.js

```diff
@@ -0,0 +1,79 @@
+import { storage, pubsub as gcpPubsub } from '../firebase.js';
+import { pubsub } from '../constants.js';
+
+/**
+ * Upload an image to the specified bucket
+ * @param {string} userId userId
+ * @param {string} experimentId experimentId
+ * @param {Blob} image image file to be uploaded
+ * @param {string} bucket destination bucket
+ *
+ * @return {JSON}
+ */
+export const uploadImageToStorage = async (
+  userId, experimentId, image, bucket) => {
+  const imagePath = `${userId}-${experimentId}/${image.name}`;
+
+  const storageRef = storage.ref();
+  const rawImageBucketRef = storageRef.child(bucket);
+  const newImageRef = rawImageBucketRef.child(imagePath);
+  newImageRef.put(image).then((snapshot) => {
+    return {
+      status: '201',
+      error: null,
+    };
+  }).catch((error) => {
+    return {
+      status: '500',
+      data: null,
+      error,
+    };
+  });
+};
+
+
+/**
+ * Subscribes to pubsub to signal stimuli generation completion
+ */
+export const observeStimuliCompletion = async () => {
+  gcpPubsub.topic(pubsub.SIE_RESULT).onPublish((msg) => {
+    // Decode the PubSub Message body.
+    const msgBody = msg.data ?
+      Buffer.from(message.data, 'base64').toString() : null;
+    // eslint-disable-next-line no-unused-vars
+    const participantId = msg.attributes.participant_id;
+    // eslint-disable-next-line no-unused-vars
+    const experimentId = msg.attributes.experiment_id;
+
+    if (msgBody === 'completed') {
```

Comment:
RFC: what should we do here?

Reply:
I am not part of the reviewers for this PR but I figured I could provide some ideas.

In my opinion, since this is an "API", I think returning something would be a good solution rather than using a hook. But a disclaimer here is that I have basically 0 experience with React hooks so I could be wrong.

In terms of that to return, I think there could be a few options, for example:

1. Return a success code such as 200 (OK) for when completed and 204 (NO_RESPONSE) for when still not completed.
2. Return some message that looks like: `{message: "completed", fileUrls: [<urls go here>]}` for when completed and `{message: "not_completed", fileUrls: []}` for when still not completed.

Note that for option #2, the `fileUrls` list doesn't really need to be part of the response of the `observeStimuliCompletion()` API. I just put it in there as an idea that could save us an additional call to `getFileUrlsFromBucket()`.

Reply:
yea I thought about that but I'm not sure because this is basically subscribing to a pubsub topic so technically it is not returning anything, more like execute this action if you receive a message kinda thing. In this particular case, we should only let participants start the image selection phase once they have finished all prior steps && a completion message above is received.

Reply:
Yeah so just a little clarification,  this method is supposed to be an **ongoing active listener**, but not a function to be called when the frontend wants to get processed images.

so once it has got the "completed" message, it will 
- fetch stimuli image URLs from the cloud storage bucket(SIE_STIMULI_IMGS)
- store those image URLs into the firestore collection ("ProcessedImages" collection)

More like an automation process running.

For frontend to get the stimuli images: I used the imageService.js to do CRUD operation for the "ProcessImages" collection and the frontend can call the getProcessedImages  to fetch those URLs from firestore directly.

Reply:
@cmutuc25 fyi
@theodore3131 why do we need to store the stimuli in a firestore collection?

Reply:
So if the stimuli images url is only used at the next step of the image selection task, we can definitely just use react hook function setState() to store them for use next step. I was assuming that if we may use it in any future steps, we can easily retrieve them from firestore collection. But if not we can just simply go with react hook function.

Reply:
yea I think we can just cache them in browser

Reply:
Oh I also realized that since this is a background listening function, we also need to check the participantID and experimentID is align with the current one to get the correction message.  

So I think we still need the frontend call this function passing the participantID and experimentID. I can push the commit to this branch if you want.

Reply:
so I did it a little differently here. I don't want every user to have to check every message, rather there should be a way for us to parse the message and signal the corresponding participant to start the survey. Let's see how the frontend wants to do this.

Reply:
Sure, so I guess this is also one of the reason why I use the firestore. We can use the firestore to store the data into the corresponding participant without checking the message to align with the current one.

However, this also brings another concern:

if two users doing it at the same time, there will be two subscribers,
they will get the same message, say 'user1-001-complete' 
then via this way, it will store the same URLs **twice** into the firestore as two users call this.

But yeah, let's talk with the frontend team to decide how to deal with this.

Reply:
Oh oops, I wasn't that was more of an ongoing listener. Thanks for explaining that!

So, how exactly would the React hook alternative work? I am visualizing that once the `observeStimuliCompletion()` gets the 'completed' message, then it could update the state or somehow let the frontend component know that the stimuli images are ready and when the frontend gets that cue then we go ahead and use the `getFileUrlsFromBucket()` API to fetch the images and go about the experiment?

If it's cumbersome to answer my questions it's all good, I can just ask other people in the FE team later to explain this to me 👍

Reply:
yea that's exactly how I'm envisioning it, not so sure about the details yet